### PR TITLE
fix: fall back to err.Error() in classifyRestoreError when stderr is empty

### LINKED ISSUE #3797: CI Doctor - CI Failure Investigation - Run #9811: `restore --existing-resource-policy=none` exits with code 1 due to Service ClusterIP "alre
[Content truncated due to length]
## Summary

The **CI - KSail** workflow (run [#9811](https://github.com/devantler-tech/ksail/actions/runs/24111862739)) failed during the Vanilla+Docker system test because `ksail cluster restore --existing-resource-policy=none` exits with code 1 even though all failures are due to resources already existing in the live cluster.

## Failure Details

- **Run**: [24111862739](https://github.com/devantler-tech/ksail/actions/runs/24111862739)
- **Run #**: 9811
- **Commit**: `aebb2c0b45c7525c9c078fbf26fc277232912afb`
- **Trigger**: `merge_group` (PR #3792)
- **Failed Job**: `🧪 System Test (Vanilla, Docker, false, --gitops-engine Flux --local-registry ...)`
- **Failed Step**: `ksail cluster restore --input /tmp/cluster-backup.tar.gz --existing-resource-policy=none`

### The Bug: `classifyRestoreError` doesn't handle Service ClusterIP "already allocated" errors

In `pkg/cli/cmd/cluster/cluster.go`, the `classifyRestoreError` function suppresses errors only when **all** stderr lines contain `"already exists"`:

````go
// pkg/cli/cmd/cluster/cluster.go ~line 4355
if flags.existingResourcePolicy == resourcePolicyNone &&
    allLinesContain(stderr, "already exists") {
    return nil
}
```

When restoring Services to a live cluster, Kubernetes returns a **different error pattern**:

```
Error from server (Invalid): error when creating "...": Service "kubernetes" is invalid:
spec.clusterIPs: Invalid value: ["10.96.0.1"]: failed to allocate IP 10.96.0.1:
provided IP is already allocated
```

This error contains **`"already allocated"`**, not `"already exists"`. So `allLinesContain(stderr, "already exists")` returns `false`, the error is not suppressed, and it gets accumulated into `restoreErrors`. At the end of `restoreResources`, since `len(restoreErrors) > 0`, the function returns `ErrRestoreFailed`, causing exit code 1.

### Error Flow

1. `restoreResourceFile` runs `kubectl create -f services.yaml`
2. Services already exist → Kubernetes returns `Invalid` (IP already allocated)
3. `classifyRestoreError(err, stderr, flags)` → `allLinesContain(stderr, "already exists")` returns `false` (Service errors say "already allocated")
4. Error string is returned and appended to `restoreErrors`
5. `restoreResources` returns `ErrRestoreFailed`
6. `runRestore` returns the error → exit code 1

## Failed Jobs and Errors

```
Warning: failed to restore services.yaml: kubectl failed:
  Error from server (Invalid): Service "kubernetes" is invalid:
    spec.clusterIPs: Invalid value: ["10.96.0.1"]:
    failed to allocate IP 10.96.0.1: provided IP is already allocated
  Error from server (Invalid): Service "kube-dns" is invalid: ...
  Error from server (Invalid): Service "source-controller" is invalid: ...
  ...
##[error]Process completed with exit code 1.
````

## Investigation Findings

The `--existing-resource-policy=none` is documented as "skip existing resources". The intent is that when restoring to a live cluster, all conflicts should be treated as benign and the command should exit 0.

However, the check only looks for `"already exists"` in stderr. Kubernetes returns at least two distinct error patterns for "resource already exists":
1. **Duplicate create**: `Error from server (AlreadyExists): ... "X" already exists`
2. **Service ClusterIP conflict**: `Error from server (Invalid): ... failed to allocate IP X: provided IP is already allocated`

Both represent the same semantic situation (resource/slot already in use), but only pattern 1 is currently suppressed.

## Recommended Actions

- [ ] **Fix `classifyRestoreError`** in `pkg/cli/cmd/cluster/cluster.go` to also suppress `"already allocated"` errors when policy is `none`:

````go
func classifyRestoreError(err error, stderr string, flags *restoreFlags) error {
    if err == nil {
        return nil
    }

    if flags.existingResourcePolicy == resourcePolicyNone &&
        allLinesAlreadyExistOrAllocated(stderr) {
        return nil
    }

    if stderr != "" {
        return fmt.Errorf("kubectl failed: %w (output: %s)", err, stderr)
    }

    return fmt.Errorf("kubectl failed: %w", err)
}

// allLinesAlreadyExistOrAllocated returns true if every non-empty line
// in output describes an already-existing or already-allocated resource.
func allLinesAlreadyExistOrAllocated(output string) bool {
    hasNonEmptyLine := false
    for line := range strings.SplitSeq(output, "\n") {
        trimmed := strings.TrimSpace(line)
        if trimmed == "" {
            continue
        }
        hasNonEmptyLine = true
        if !strings.Contains(trimmed, "already exists") &&
            !strings.Contains(trimmed, "already allocated") {
            return false
        }
    }
    return hasNonEmptyLine
}
```

- [ ] **Add unit test** for `classifyRestoreError` covering the Service IP "already allocated" case
- [ ] Update the E2E test step comment: the `--existing-resource-policy=none` test correctly expects exit 0 (no `|| true` needed once the fix is applied)

## Prevention Strategies

- When introducing new `--existing-resource-policy` behavior, test against a **live** (non-empty) cluster, not a fresh one
- The `classifyRestoreError` logic should be tested with real Kubernetes API error messages from `kubectl create`
- Consider using `kubectl apply --dry-run=server` to detect conflicts more uniformly, instead of parsing error message strings

## AI Team Self-Improvement

Add to `copilot-instructions.md` for AI coding agents:

```
## Restore Error Classification (`pkg/cli/cmd/cluster/cluster.go`)

When classifying whether a `kubectl create` error is "benign" (resource already exists)
for `--existing-resource-policy=none`, check for BOTH Kubernetes error patterns:

1. `"already exists"` — standard AlreadyExists (most resource types)
2. `"already allocated"` — Service ClusterIP conflicts (spec.clusterIPs invalid)

Using `allLinesContain(stderr, "already exists")` alone will miss Service IP conflicts
and cause `restore --existing-resource-policy=none` to return exit code 1 on live clusters.
````

## Historical Context

- **Run #9802** (issue #3794): `tar | grep -q` broken pipe in K3s namespace-filtered backup test (different step, same PR #3792)
- **Run #9811** (this issue): `restore --existing-resource-policy=none` exits 1 due to Service ClusterIP error message mismatch
- Both failures stem from new E2E test coverage added in PR #3792

> Generated by [CI Doctor](https://github.com/devantler-tech/ksail/actions/runs/24112022452/agentic_workflow) · ● 3.4M · [◷](https://github.com/search?q=repo%3Adevantler-tech%2Fksail+is%3Aissue+%22gh-aw-workflow-call-id%3A+devantler-tech%2Fksail%2Fci-doctor%22&type=issues)
>
> To install this [agentic workflow](https://github.com/githubnext/agentics/blob/7c7feb61a52b662eb2089aa2945588b7a200d404/workflows/ci-doctor.md), run
> ```
> gh aw add githubnext/agentics/workflows/ci-doctor.md@7c7feb61a52b662eb2089aa2945588b7a200d404
> ```

### PR DESCRIPTION
**Pull request by devantler**
## Summary

Fixes #3797

When using `--existing-resource-policy=none`, some resource types (DaemonSet, Job) route "AlreadyExists" errors through `BehaviorOnFatal` (into `err.Error()`) instead of stderr. When stderr is empty, `allLinesContain("", "already exists")` returns `false` (no non-empty lines), so the benign error was not suppressed — causing `ksail cluster restore` to exit with code 1.

## Changes

- **`pkg/cli/cmd/cluster/cluster.go`**: In `classifyRestoreError`, fall back to checking `err.Error()` when stderr is empty
- **`pkg/cli/cmd/cluster/export_test.go`**: Export `classifyRestoreError` for testing
- **`pkg/cli/cmd/cluster/cluster_test.go`**: Add 7 test cases covering the fallback path (empty stderr with err containing "already exists", mixed errors, policy=update, etc.)

## Root cause

```
daemonsets.yaml: kubectl failed: Error from server (AlreadyExists): error when creating "...": daemonsets.apps "svclb-traefik-*" already exists
jobs.yaml: kubectl failed: Error from server (AlreadyExists): error when creating "...": jobs.batch "helm-install-traefik" already exists
```

The error message traveled through `BehaviorOnFatal` → `err.Error()`, but `classifyRestoreError` only inspected `stderr`.

### CODE REVIEW

File: pkg/cli/cmd/cluster/cluster.go [hunk 1]
```diff
@@ -4352,9 +4352,18 @@ func classifyRestoreError(err error, stderr string, flags *restoreFlags) error {
...
-		allLinesContain(stderr, "already exists") {
-		return nil
+	if flags.existingResourcePolicy == resourcePolicyNone {
+		// Some resource types (e.g. DaemonSet, Job) route
+		// "AlreadyExists" through BehaviorOnFatal instead of stderr.
+		// Fall back to err.Error() when stderr is empty.
+		source := stderr
+		if source == "" {
+			source = err.Error()
+		}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/cluster/cluster.go:4363]: Good catch! Updated to use `strings.TrimSpace(stderr) == ""` so whitespace-only stderr correctly triggers the fallback to `err.Error()`.

File: pkg/cli/cmd/cluster/cluster.go [hunk 2]
```diff
@@ -4352,9 +4352,18 @@ func classifyRestoreError(err error, stderr string, flags *restoreFlags) error {
...
+
+		if allLinesContain(source, "already exists") {
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/cluster/cluster.go:4367]: Thanks for raising this. The "already allocated" (ClusterIP conflict) is a different error  it means the resource does not exist but its requested IP is taken by another Service. Treating it as benign could silently drop resources from the restore. The scope of this PR is the specific bug from #3797 where "already exists" errors routed through BehaviorOnFatal were not suppressed. Expanding to other conflict patterns would be a separate enhancement.class

File: pkg/cli/cmd/cluster/cluster_test.go [hunk 3]
```diff
@@ -6141,3 +6141,83 @@ func TestDisplayComponents_WithState(t *testing.T) {
...
+			expectNil: true,
+		},
+		{
+			name:      "already exists in err.Error() with empty stderr",
+			err:       errors.New("Error from server (AlreadyExists): daemonsets.apps \"svclb-traefik\" already exists"),
+			stderr:    "",
+			policy:    "none",
+			expectNil: true,
+		},
+		{
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/cluster/cluster_test.go:6206]: Good suggestion! Added the whitespace-only stderr test case alongside the TrimSpace fix.

File: pkg/cli/cmd/cluster/cluster_test.go [hunk 4]
```diff
@@ -6141,3 +6141,102 @@ func TestDisplayComponents_WithState(t *testing.T) {
...
+	}
+
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
+			t.Parallel()
+
+			result := cluster.ExportClassifyRestoreError(
+				testCase.err, testCase.stderr, testCase.policy,
+			)
+			if testCase.expectNil {
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/cluster/cluster_test.go:6243]: This is a false  Go 1.22+ changed range loop semantics so each iteration gets its own copy of the variable. This project uses Go 1.26.1 (see go.mod). All 16 existing parallel test loops in this file use `testCase` directly without capturing, consistent with this pattern.positive
